corehq/util/couch.py

```diff
@@ -33,3 +36,130 @@ def categorize_bulk_save_errors(error):
+        self.to_save = []
+
+    def commit_delete(self):
+        success_ids = self.commit(self.db.bulk_delete, self.to_delete)
+        self.deleted_ids.extend(success_ids)
+        self.to_delete = []
+
+    def save(self, doc):
+        self.to_save.append(doc)
+        if len(self.to_save) >= self.chunksize:
+            self.commit_save()
+
+    def delete(self, doc):
+        self.to_delete.append(doc)
+        if len(self.to_delete) >= self.chunksize:
+            self.commit_delete()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.to_save:
+            self.commit_save()
+        if self.to_delete:
+            self.commit_delete()
+
+
+def iter_update(db, fn, ids):
+    """
+    Map `fn` over every doc in `db` matching `ids`
+
+    `fn` should accept a json dict from couch and return one of:
+     * a json dict (or wrapped couch document) - bulk save that document
+     * the string "DELETE" - bulk delete that document
+     * None - pass
+
+    iter_update returns an object with the following properties:
+    saved_ids, deleted_ids, ignored_ids,
+
+    Ex: mark dimagi users as cool, delete the Canadians, and ignore the rest
+
+        def mark_cool(user_dict):
+            user = CommCareUser.wrap(user_dict)
+            if user.is_dimagi:
+                user.is_cool = True
+                return user
+            if user.language == "Canadian":
+                return "DELETE"
+
+        iter_update(CommCareUser.get_db(), mark_cool, all_user_ids)
+
+    This looks up and saves docs in chunks and is intended for large changes
+    such as a migration.  If an id is not found, it is skipped.  In the
+    event of a BulkSaveError, it will re-process the unsaved documents.
+    Wrapping is optional, and this function will unwrap if needed.
+    """
+    IterResults = namedtuple('IterResults', ['ignored_ids', 'not_found_ids',
+                                             'deleted_ids', 'updated_ids'])
+    results = IterResults(set(), set(), set(), set())
+
+    def _iter_update(doc_ids):
+        with IterDB(db, chunksize=100) as iter_db:
+            for chunk in chunked(set(doc_ids), 100):
+                for res in get_docs(db, keys=chunk):
+                    raw_doc = res.get('doc')
+                    doc_id = res.get('id', None)
+                    if not raw_doc or not doc_id:
+                        results.not_found_ids.add(res['key'])
+                    else:
+                        doc = fn(raw_doc)
+                        if isinstance(doc, dict):
+                            iter_db.save(doc)
+                        elif doc == "DELETE":
+                            iter_db.delete(raw_doc)
+                        else:
+                            results.ignored_ids.add(doc_id)
+
+        results.updated_ids.update(iter_db.saved_ids)
+        results.deleted_ids.update(iter_db.deleted_ids)
+
+        if iter_db.error_ids:
+            _iter_update(iter_db.error_ids)
```

Comment:
I just added the automatic retry support, but I disagree that the default should be to not retry.

For `IterDB`, the caller is providing the documents and telling when to save, so in that case the responsibility clearly lies on them to handle errors.

`iter_update`, on the other hand, is given a function to be called on every doc in a list of ids, and the caller has no knowledge of the docs' states.  In this case then, this function should be applicable regardless of the doc's state, and any conflicts are just the result of bad timing between the get and the write, both of which happen in `iter_update`.

I added a max_retries param, which the caller can set to `0` to avoid retries, does that sound reasonable?